### PR TITLE
fix: Enforcing a feasible solution for a `call-center` flaky test

### LINKED ISSUE #30: Flaky SolverServiceTest test in Call Center quickstart
```
2023-06-19T13:32:32.5305270Z [ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.478 s <<< FAILURE! - in org.acme.callcenter.solver.SolverServiceTest
2023-06-19T13:32:32.5306210Z [ERROR] org.acme.callcenter.solver.SolverServiceTest.removeCall  Time elapsed: 0.021 s  <<< FAILURE!
2023-06-19T13:32:32.5306850Z java.lang.AssertionError: 
2023-06-19T13:32:32.5307200Z 
2023-06-19T13:32:32.5307490Z Expecting RegularEnumSet:
2023-06-19T13:32:32.5308000Z   [ENGLISH, LIFE_INSURANCE, PROPERTY_INSURANCE]
2023-06-19T13:32:32.5309170Z to contain:
2023-06-19T13:32:32.5309580Z   [ENGLISH, CAR_INSURANCE]
2023-06-19T13:32:32.5310050Z but could not find the following element(s):
2023-06-19T13:32:32.5310340Z   [CAR_INSURANCE]
2023-06-19T13:32:32.5310480Z 
2023-06-19T13:32:32.5310710Z 	at org.acme.callcenter.solver.SolverServiceTest.removeCall(SolverServiceTest.java:90)
```

### PR DESCRIPTION
**Pull request by zepfred**
This PR attempts to fix a flaky test (`org.acme.callcenter.solver.SolverServiceTest#removeCall`) with an unclear failure reason. We have noticed that the Solver sometimes fails to find a feasible solution when it returns the best solution. Therefore, we have made changes to ensure the system will wait for the Solver to find a feasible solution for up to 30 seconds before returning the best solution.

Issue #30

### CODE REVIEW

**Comment by triceo** [comment]
@rsynek Please review Fred's PR. If my memory serves, you wrote this quickstart, so you should know more about it than I do.

File: use-cases/call-center/src/test/java/org/acme/callcenter/solver/retry/RetryTestExtension.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        public void incFailures() {
+            failures++;
+        }
+
+        @Override
+        public boolean hasNext() {
+            boolean hadSuccessful = executions > failures;
+            // Stop when there in one successful or the max attempts is exhausted
```
Comment (rsynek) [inline use-cases/call-center/src/test/java/org/acme/callcenter/solver/retry/RetryTestExtension.java]:
```suggestion
            // Stop when there is one successful or the max attempts is exhausted
```

File: use-cases/call-center/src/test/java/org/acme/callcenter/solver/SolverServiceTest.java [hunk 2]
```diff
@@ -77,6 +76,7 @@ void prolongCall() {
 
     @Test
     @Timeout(60)
+    @Retry(3)
```
Comment (rsynek) [inline use-cases/call-center/src/test/java/org/acme/callcenter/solver/SolverServiceTest.java]: While this approach lowers the risk of the test failing again, it does not fix the root cause.

This is a not a problem with a random failure due to e.g. network latency, which would justify the retries.

Suggestion: try reproducing the issue by running the test in a loop (perhaps with some very short random pauses) and see if it shows up.

**Review by rsynek — COMMENTED** [review]
The retried test looks ok and will make the issue disappear from our radar. If a quick fix is what we are looking for, let's merge it.

However, the root cause is still escaping our attention and if a user copies this QS as a base of their project, the issue might cause them harm.

File: use-cases/call-center/src/test/java/org/acme/callcenter/solver/SolverServiceTest.java (hunk 2 above)
Comment (zepfred) [reply use-cases/call-center/src/test/java/org/acme/callcenter/solver/SolverServiceTest.java]: Right. I couldn't reproduce the issue, and your suggestion makes sense. Let me give it a try.

**Comment by zepfred** [comment]
@rsynek, I've updated the PR description to match our discussed strategy.

**Review by rsynek — APPROVED** [review]
Thanks for the fix; let's merge it and see if the issue ever shows up again.

**Pull request merged**